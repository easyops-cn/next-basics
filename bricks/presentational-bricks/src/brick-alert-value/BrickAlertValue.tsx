import React from "react";
import { get, find } from "lodash";

import { MonitorModels } from "@next-sdk/monitor-sdk";

interface BrickAlertValueProps {
  alertEvent: MonitorModels.ModelAlertEvent;
}

function computeAlertValue(alertEvent: MonitorModels.ModelAlertEvent): string {
  const unit = get(alertEvent, "alert_conditions.unit", "");
  const comparator = get(
    find(get(alertEvent, "alert_conditions.comparators", []), {
      level: alertEvent.level
    }),
    "type"
  );

  const compareMapping: any = {
    bigger_than: "↑",
    smaller_than: "↓"
  };
  const map: string = compareMapping[comparator]
    ? compareMapping[comparator]
    : "";

  if (typeof alertEvent.value !== "number") {
    return alertEvent.value;
  }

  return (
    (Number.isInteger(alertEvent.value)
      ? alertEvent.value
      : alertEvent.value.toFixed(2)) +
    unit +
    " " +
    map
  );
}

export function BrickAlertValue(
  props: BrickAlertValueProps
): React.ReactElement {
  const alertValue = computeAlertValue(props.alertEvent);

  return <span>{alertValue}</span>;
}
