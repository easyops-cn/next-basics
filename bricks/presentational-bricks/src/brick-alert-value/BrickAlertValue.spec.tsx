import React from "react";
import { shallow } from "enzyme";
import { BrickAlertValue } from "./BrickAlertValue";

describe("BrickAlertValue", () => {
  it("should work, with bigger", () => {
    const alertEvent: any = {
      alert_conditions: {
        alert_type: "host",
        comparators: [
          {
            threshold: 95,
            type: "bigger_than",
            level: 1
          }
        ],
        alert_sub_type: "host",
        attr_id: "host.disk.max_used_percent",
        alert_table: "host",
        unit: "%"
      },
      value: 98,
      level: 1
    };

    const wrapper = shallow(<BrickAlertValue alertEvent={alertEvent} />);
    expect(wrapper).toMatchSnapshot();
  });

  it("should work, without bigger or smaller", () => {
    const alertEvent: any = {
      alert_conditions: {
        alert_type: "host",
        alert_sub_type: "host",
        attr_id: "host.disk.max_used_percent",
        alert_table: "host",
        unit: "%"
      },
      value: 98.9987,
      level: 1
    };

    const wrapper = shallow(<BrickAlertValue alertEvent={alertEvent} />);
    expect(wrapper).toMatchSnapshot();
  });

  it("should work, value is not number", () => {
    const alertEvent: any = {
      alert_conditions: {
        alert_type: "host",
        alert_sub_type: "host",
        attr_id: "host.disk.max_used_percent",
        alert_table: "host",
        unit: "%"
      },
      value: "192.168.100.162",
      level: 1
    };

    const wrapper = shallow(<BrickAlertValue alertEvent={alertEvent} />);
    expect(wrapper).toMatchSnapshot();
  });
});
