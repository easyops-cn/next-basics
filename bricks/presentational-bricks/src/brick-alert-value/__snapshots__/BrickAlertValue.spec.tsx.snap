// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BrickAlertValue should work, value is not number 1`] = `
<span>
  192.168.100.162
</span>
`;

exports[`BrickAlertValue should work, with bigger 1`] = `
<span>
  98% â†‘
</span>
`;

exports[`BrickAlertValue should work, without bigger or smaller 1`] = `
<span>
  99.00% 
</span>
`;
